### PR TITLE
test: add tests for CheckAssociatedObjects

### PR DESCRIPTION
**Pull request by adriantam**
## Description
Adding unit test for buildCheckAssociatedObjects.  This will serve as building block to test consumeUsersets for graph/check.  In addition, update buildCheckAssociatedObjects so that it is not a method for the local checker 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: internal/graph/check_test.go [hunk 1]
```diff
@@ -2998,3 +3000,329 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			ts, err := typesystem.NewAndValidate(context.Background(), tt.model)
+			require.NoError(t, err)
+
+			contextStruct, err := structpb.NewStruct(tt.context)
+			require.NoError(t, err)
+
+			iterFunc := buildTupleKeyConditionFilter(context.Background(), contextStruct, ts)
```
Comment (miparnisari) [inline internal/graph/check_test.go:3095]: Not for this PR, but: i am starting to feel like this file will become huge with tests. That is not a bad thing, but here we are testing a function that is not strictly related to Check and could be used in other APIs.

I feel like we need a `check_utils` file or something where all the auxiliary functions could live.

File: internal/graph/check_test.go [hunk 2]
```diff
@@ -2998,3 +3000,329 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+
+				type document
+					relations
+						define can_view: [user]`),
+			context:      map[string]interface{}{},
+			conditionMet: false,
+			expectedErr:  condition.NewEvaluationError("correct_ip", fmt.Errorf("condition was not found")),
+		},
+		{
+			name:     "condition_missing_parameter",
```
Comment (miparnisari) [inline internal/graph/check_test.go:3046]: given this test name don't you need to define the condition in the model? otherwise this test is the same as the previous one

File: internal/graph/check_test.go [hunk 3]
```diff
@@ -2998,3 +3000,329 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+	}
+}
+
+func TestBuildCheckAssociatedObjects(t *testing.T) {
+	tests := []struct {
+		name                         string
+		model                        *openfgav1.AuthorizationModel
+		tuples                       []*openfgav1.TupleKey
+		context                      map[string]interface{}
+		dsError                      bool
```
Comment (miparnisari) [inline internal/graph/check_test.go:3113]: this is always false

File: internal/graph/check_test.go (hunk 1 above)
Comment (adriantam) [reply internal/graph/check_test.go:3095]: Absolutely, we can move the function that are not direct local checkers into this check_utils function

Comment (adriantam) [reply internal/graph/check_test.go:3095]: I probably want to do this after the next PR (where it is nothing but move functions out into the util file)

File: internal/graph/check_test.go (hunk 2 above)
Comment (adriantam) [reply internal/graph/check_test.go:3046]: remove redundant call

File: internal/graph/check_test.go (hunk 3 above)
Comment (adriantam) [reply internal/graph/check_test.go]: add test on dsError case

File: internal/graph/check_test.go (hunk 1 above)
Comment (senojj) [reply internal/graph/check_test.go:3079]: perhaps a new package is what we should aim for? something like `checkhelper`?

Comment (senojj) [reply internal/graph/check_test.go:3079]: however a more meaningful package would be ideal

Comment (adriantam) [reply internal/graph/check_test.go:3079]: We can have something like checkUtil package and put all the util function there.

Comment (miparnisari) [reply internal/graph/check_test.go:3079]: `buildTupleKeyConditionFilter` can be re-used in Check, ListObjects and Listusers, and we can fix this: https://github.com/openfga/openfga/issues/1567

I'd move to the `storage` folder, maybe inside `iterators`

File: internal/graph/check_test.go [hunk 4]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+			for _, objectID := range tt.objectIDs {
+				objectIDs.Add(objectID)
+			}
+
+			associatedFunc := buildCheckAssociatedObjects(&ResolveCheckRequest{
+				StoreID:              ulid.Make().String(),
+				AuthorizationModelID: ulid.Make().String(),
+				TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:maria"),
+				RequestMetadata:      NewCheckRequestMetadata(20),
+			}, "group#member", objectIDs)
```
Comment (miparnisari) [inline internal/graph/check_test.go:3309]: Not for this PR: we can simplify the signature of this function:

```go
func buildCheckAssociatedObjects(ctx context.Context, req *ResolveCheckRequest, objectRel string, objectIDs storage.SortedSet) (*ResolveCheckResponse, error) {
```

File: internal/graph/check_test.go [hunk 5]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("group:1", "member", "user:maria"),
+				tuple.NewTupleKey("group:2", "member", "user:maria"),
+				tuple.NewTupleKey("group:3", "member", "user:maria"),
+				tuple.NewTupleKey("group:4", "member", "user:maria"),
+			},
+			dsError: nil,
+			objectIDs: []string{
+				"2", "3",
+			},
```
Comment (miparnisari) [inline internal/graph/check_test.go:3173]: Nitpick: you don't need 4 tuples to create a truthy response, just one tuple should be enough. Or are we also testing the behavior of `tupleIDInSortedSet`, which already has its own tests? 😄

File: internal/graph/check_test.go [hunk 6]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+			t.Cleanup(ctrl.Finish)
+			ds := mocks.NewMockRelationshipTupleReader(ctrl)
+			tuples := make([]*openfgav1.Tuple, len(tt.tuples))
+			for i, tuple := range tt.tuples {
+				tuples[i] = &openfgav1.Tuple{
+					Key: tuple,
+				}
+			}
+			dsIter := storage.NewStaticTupleIterator(tuples)
+			ds.EXPECT().ReadStartingWithUser(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(dsIter, tt.dsError)
```
Comment (miparnisari) [inline internal/graph/check_test.go:3321]:
```suggestion
			ds.EXPECT().ReadStartingWithUser(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
				Times(1).
				Return(storage.NewStaticTupleIterator(tt.tuples), tt.dsError)
```

and then change the type of `tt.tuples` to `[]*openfgav1.Tuple` :)

File: internal/graph/check_test.go [hunk 7]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+			},
+			dsError: nil,
+			objectIDs: []string{
+				"2", "3",
+			},
+			context: map[string]interface{}{
+				"x": 1,
+			},
+			expectedResolveCheckResponse: &ResolveCheckResponse{
+				Allowed: false,
```
Comment (miparnisari) [inline internal/graph/check_test.go:3286]: Why is this false? The condition is satisfied for every tuple and there is an intersection for group:2 and group:3

File: internal/graph/check_test.go [hunk 8]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+
+				condition condX(x: int) {
+					x < 100
+				}
+`),
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKeyWithCondition("group:1", "viewer", "user:maria", "condX", nil),
+				tuple.NewTupleKeyWithCondition("group:2", "viewer", "user:maria", "condX", nil),
+				tuple.NewTupleKeyWithCondition("group:3", "viewer", "user:maria", "condX", nil),
+				tuple.NewTupleKeyWithCondition("group:4", "viewer", "user:maria", "condX", nil),
```
Comment (miparnisari) [inline internal/graph/check_test.go:3237]: these tuples are not valid according to the model

File: internal/graph/check_test.go [hunk 9]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+
+			objectIDs := storage.NewSortedSet()
+			for _, objectID := range tt.objectIDs {
+				objectIDs.Add(objectID)
+			}
+
+			associatedFunc := buildCheckAssociatedObjects(&ResolveCheckRequest{
+				StoreID:              ulid.Make().String(),
+				AuthorizationModelID: ulid.Make().String(),
+				TupleKey:             tuple.NewTupleKey("document:1", "viewer", "user:maria"),
```
Comment (miparnisari) [inline internal/graph/check_test.go:3307]: don't we need to test different inputs here too?

File: internal/graph/check_test.go [hunk 10]
```diff
@@ -2998,3 +3000,338 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+		model                        *openfgav1.AuthorizationModel
+		tuples                       []*openfgav1.TupleKey
+		context                      map[string]interface{}
+		dsError                      error
+		objectIDs                    []string
+		expectedError                bool
+		expectedResolveCheckResponse *ResolveCheckResponse
+	}{
+		{
+			name: "empty_iterator",
```
Comment (miparnisari) [inline internal/graph/check_test.go:3103]: should we also have `empty_object_ids_list`?

File: internal/graph/check_test.go (hunk 7 above)
Comment (adriantam) [reply internal/graph/check_test.go:3286]: This should be invalid test as the relation does not exist.  Let me nuke this.

File: internal/graph/check_test.go (hunk 5 above)
Comment (adriantam) [reply internal/graph/check_test.go:3173]: Will simplify the test - tupleIDInSortedSet already has its own tests

File: internal/graph/check_test.go (hunk 9 above)
Comment (adriantam) [reply internal/graph/check_test.go:3307]: Nothing in this top level function should behave differently according with various model / request.  These various combination should be in the inner function such as iteratorReadStartingFromUser / NewConditionsFilteredTupleKeyIterator / tupleIDInSortedSet.  These combo should be tested in the inner function (which is the whole point of breaking them into smaller piece so we don't have a large combo in the higher up function).  If there are specific thing we want to test, we can add them here.  I will add a TTU here for sanity for now.

**Review by miparnisari — APPROVED** [review]

File: internal/graph/check_test.go [hunk 11]
```diff
@@ -2998,3 +3000,373 @@ func TestIteratorReadStartingFromUser(t *testing.T) {
...
+
+				type document
+					relations
+						define can_view: [user]`),
+			context:      map[string]interface{}{},
+			conditionMet: true,
+			expectedErr:  nil,
+		},
+		{
+			name:     "condition_not_met",
```
Comment (miparnisari) [inline internal/graph/check_test.go:3030]: can you add the case where the condition is met?

Comment (adriantam) [reply internal/graph/check_test.go:3030]: Done

File: internal/graph/check.go [hunk 12]
```diff
@@ -740,47 +740,45 @@ func iteratorReadStartingFromUser(ctx context.Context,
 		}, opts)
 }
 
-func (c *LocalChecker) buildCheckAssociatedObjects(req *ResolveCheckRequest, objectRel string, objectIDs storage.SortedSet) CheckHandlerFunc {
-	return func(ctx context.Context) (*ResolveCheckResponse, error) {
-		ctx, span := tracer.Start(ctx, "checkAssociatedObjects")
-		defer span.End()
+func buildCheckAssociatedObjects(ctx context.Context, req *ResolveCheckRequest, objectRel string, objectIDs storage.SortedSet) (*ResolveCheckResponse, error) {
```
Comment (elbuo8) [inline internal/graph/check.go:743]:
```suggestion
func checkAssociatedObjects(ctx context.Context, req *ResolveCheckRequest, objectRel string, objectIDs storage.SortedSet) (*ResolveCheckResponse, error) {
```

Doesn't make sense for it to be called `build` anymore since its not returning a function anymore.

**Review by miparnisari — APPROVED** [review]
